pkg/cloud/cloud.go

```diff
@@ -765,27 +768,130 @@ func (c *cloud) waitForVolume(ctx context.Context, volumeID string) error {
 	return err
 }
 
-// Helper function for describeVolume callers. Tries to retype given error to AWS error
-// and returns true in case the AWS error is "InvalidVolume.NotFound", false otherwise
-func isAWSErrorVolumeNotFound(err error) bool {
+// isAWSError verifies if a given error is an AWS error with the given code.
+func isAWSError(err error, code string) bool {
 	if awsError, ok := err.(awserr.Error); ok {
 		// https://docs.aws.amazon.com/AWSEC2/latest/APIReference/errors-overview.html
-		if awsError.Code() == "InvalidVolume.NotFound" {
+		if awsError.Code() == code {
 			return true
 		}
 	}
 	return false
 }
 
+func isAWSErrorIncorrectModification(err error) bool {
+	return isAWSError(err, "IncorrectModificationState")
+}
+
+// Helper function for describeVolume callers. Tries to retype given error to AWS error
+// and returns true in case the AWS error is "InvalidVolume.NotFound", false otherwise
+func isAWSErrorVolumeNotFound(err error) bool {
+	return isAWSError(err, "InvalidVolume.NotFound")
+}
+
 // Helper function for describeSnapshot callers. Tries to retype given error to AWS error
 // and returns true in case the AWS error is "InvalidSnapshot.NotFound", false otherwise
 func isAWSErrorSnapshotNotFound(err error) bool {
-	if awsError, ok := err.(awserr.Error); ok {
-		// https://docs.aws.amazon.com/AWSEC2/latest/APIReference/errors-overview.html
-		if awsError.Code() == "InvalidSnapshot.NotFound" {
-			return true
+	return isAWSError(err, "InvalidSnapshot.NotFound")
+}
+
+// ResizeDisk resizes an EBS volume in GiB increments,
+// rouding up to the next GiB if reqSize isn't an even GiB.
+// It returns the new size or a negative number if the size couldn't be determined.
+func (c *cloud) ResizeDisk(ctx context.Context, volumeID string, reqSizeBytes int64) (int64, error) {
+	// If current size isgreater or equal, than there's nothing to be done here
+	request := &ec2.DescribeVolumesInput{
+		VolumeIds: []*string{
+			aws.String(volumeID),
+		},
+	}
+	volume, err := c.getVolume(ctx, request)
+	if err != nil {
+		return -1, err
+	}
+
+	// AWS resizes in chunks of GiB (not GB)
+	reqSizeGiB := util.RoundUpGiB(reqSizeBytes)
+	oldSizeGiB := aws.Int64Value(volume.Size)
+
+	if oldSizeGiB >= reqSizeGiB {
+		err := fmt.Errorf("could not expand volume %q: current size (%d GiB) is greater or equal to requested size (%d GiB)", volumeID, oldSizeGiB, reqSizeGiB)
+		return oldSizeGiB, err
+	}
+
+	// Modify volume
+	req := &ec2.ModifyVolumeInput{
+		VolumeId: aws.String(volumeID),
+		Size:     aws.Int64(reqSizeGiB),
+	}
+
+	output, err := c.ec2.ModifyVolumeWithContext(ctx, req)
+	if err != nil {
+		if isAWSErrorIncorrectModification(err) {
+			m, err2 := c.getVolumeModification(ctx, volumeID)
+			if err2 != nil {
+				return -1, fmt.Errorf("could not get modification state of volume %q: %v", volumeID, err)
+			}
+
+			// Check for idempotency
+			state := aws.StringValue(m.ModificationState)
```

Comment:
I think we need to move the idempotency check to controller.go. That's where we handles idempotency for `CreateVolume` and `ControllerPublish` so far. And cloud.go should be a simply interface to AWS service to not handle too much other complexities.

Reply:
I removed the comment because it's misleading.